### DISCUSSION TITLE
Ghostty's shell integration works worse than iTerm2's shell integration, and it may be a potential bug

### DISCUSSION THREAD

**Discussion by Xe-no1** [ghostty-org/ghostty#3759] [Feature Requests, Ideas]
Ghostty has it's own built in shell integration, but on macOS, using zsh, it is 'buggy' when it comes to text redrawing, and the `cursor-click-to-move` feature doesn't work at all, basically, text redrawing when resizing the terminal and making it smaller (especially when there is alot of text printed to `stdout`, like when using a zsh prompt) makes the terminal look very ugly, and the `cursor-click-to-move` feature simply doesn't work at all, however, when using [iTerm2's shell integration](https://iterm2.com/documentation-shell-integration.html) and sourcing it in my `.zshrc` instead of Ghostty's default shell integration, no issues occur and everything works perfectly. This is my entire ghostty config:
```
theme = nord
font-family = JetBrainsMonoNL Nerd Font Mono
font-size = 16
window-width = 180
window-height = 51
window-padding-x = 0
window-padding-y = 0
window-padding-balance = true
confirm-close-surface = false
macos-option-as-alt = true
clipboard-paste-protection = false
unfocused-split-opacity = 0.4
copy-on-select = false
cursor-style = bar
cursor-style-blink = true
shell-integration-features = cursor,sudo,no-title
window-colorspace = display-p3
```
This is the relevant part of my `.zshrc`, which manually loads ghostty's zsh integration:
```
# Manually load ghostty zsh integration
if [[ -n $GHOSTTY_RESOURCES_DIR ]]; then
  source $GHOSTTY_RESOURCES_DIR/shell-integration/zsh/ghostty-integration
fi
```
This is the relevant script that downloads and sources iTerm2's zsh integration (which should be sourced standalone without ghostty's shell integration):
```
# Manually install and load iTerm zsh integration
if [ ! -e ~/.iterm2_shell_integration.zsh ]; then
curl -L https://iterm2.com/shell_integration/zsh -o ~/.iterm2_shell_integration.zsh
fi
source ~/.iterm2_shell_integration.zsh
```
This is a video to display an example of what happens on my machine (sorry if the quality is a bit bad I used ffmpeg to compress):

https://github.com/user-attachments/assets/812cd47e-702a-4aad-98b0-1faf18f5ba45

To recreate my problem, Ghostty should have the zsh integration feature enabled by default as well as `cursor-click-to-move = true` in the Ghostty config, and then iTerm's zsh integration is downloaded and sourced in the `.zshrc`, replacing Ghostty's zsh integration, both outcomes are compared, and iTerm's should work better, indicating that there may be a problem in Ghostty's default integration. I should also mention that I am using [powelevel10k's zsh prompt](https://github.com/romkatv/powerlevel10k). Thanks in advance for any help provided, or confirmation that this is a bug that shouldn't happen. For now I will work around this issue by using iTerm's zsh integration, but hopefully a fix in Ghostty's default one will be provided!

**Comment by Xe-no1** [comment]
***Fix by using iTerm2's integration instead of Ghostty's***
Manually iTerm's shell integration fixes it, this should be added to the `.zshrc`
```
# Manually install and load iTerm zsh integration
if [ ! -e ~/.iterm2_shell_integration.zsh ]; then
curl -L https://iterm2.com/shell_integration/zsh -o ~/.iterm2_shell_integration.zsh
fi
source ~/.iterm2_shell_integration.zsh
```

**Comment by Xe-no1** [comment]
Im reopening this discussion, because although I can just work around this by using iTerm's integration, I'm hoping to make ghostty an even better terminal for everyone, so I must reopen this in hopes of actually finding a real fix for the *builtin* shell integration, not working around it with iTerm.

**Comment by mitchellh** [comment]
We've since rewritten our Zsh integration. Is this still an issue?

**Reply by Roxiun** [reply]
~~Yes, this is still an issue~~

No it appears it is fixed. By default the shell integration doesn't seem to be detected though and I had to manually add 

```
if [[ -n $GHOSTTY_RESOURCES_DIR ]]; then
  source $GHOSTTY_RESOURCES_DIR/shell-integration/zsh/ghostty-integration
fi
```

to my `.zshrc`, possible because I might be using the deafult zsh packaged with macos

**Reply by Roxiun** [reply]
Actually on further testing it is still and issue,  particularly when resizising text using CMD + shortcut, where as iTerm2 integration doesn't have any issues
